lib/libutee/tee_api_property.c

```diff
@@ -162,290 +177,250 @@ static TEE_Result propget_get_property(TEE_PropSetHandle h, char *name,
 			return res;
 
 		if (idx < eps_len)
-			return propget_get_ext_prop(eps + idx, pv);
+			return propget_get_ext_prop(eps + idx, type, buf, len);
 		idx -= eps_len;
 
-		size = sizeof(pv->u);
 		res = utee_get_property((unsigned long)pe->prop_set, idx,
-					0, 0, &pv->u, &size, &prop_type);
+					NULL, NULL, buf, len, &prop_type);
 		if (res == TEE_ERROR_ITEM_NOT_FOUND)
 			res = TEE_ERROR_BAD_PARAMETERS;
 	}
 
-	if (res == TEE_SUCCESS)
-		pv->type = prop_type;
+	*type = prop_type;
 	return res;
 }
 
 TEE_Result TEE_GetPropertyAsString(TEE_PropSetHandle propsetOrEnumerator,
-				   char *name, char *valueBuffer,
-				   uint32_t *valueBufferLen)
+				   char *name, char *value,
+				   uint32_t *value_len)
 {
 	TEE_Result res;
-	struct prop_value pv;
 	size_t l;
-	size_t bufferlen;
+	enum user_ta_prop_type type;
+	void *tmp_buf = 0;
+	uint32_t tmp_len;
+	uint32_t uint32_val;
+	TEE_Identity *p_identity_val;
 
-	if (valueBuffer == NULL || valueBufferLen == NULL) {
+	if (!value || !value_len) {
 		res = TEE_ERROR_BAD_PARAMETERS;
-		goto err;
+		goto out;
 	}
 
-	bufferlen = *valueBufferLen;
+	tmp_len = *value_len;
+	if (tmp_len < sizeof(TEE_Identity))
+		tmp_len = sizeof(TEE_Identity);
+	tmp_buf = TEE_Malloc(tmp_len, TEE_USER_MEM_HINT_NO_FILL_ZERO);
+	if (!tmp_buf) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
 
-	res = propget_get_property(propsetOrEnumerator, name, &pv);
-	if (res != TEE_SUCCESS)
-		goto err;
+	res = propget_get_property(propsetOrEnumerator, name, &type,
+				   tmp_buf, &tmp_len);
+	if (res != TEE_SUCCESS) {
+		if (res == TEE_ERROR_SHORT_BUFFER)
+			*value_len = tmp_len;
+		goto out;
+	}
 
-	switch (pv.type) {
+	switch (type) {
 	case USER_TA_PROP_TYPE_BOOL:
-		l = strlcpy(valueBuffer, pv.u.bool_val ? "true" : "false",
-			    bufferlen);
+		uint32_val = *((uint32_t *)tmp_buf);
+		l = strlcpy(value, (uint32_val ? "true" : "false"),
```

Comment:
`strlcpy` does not count the last 0 zero termination. So I do not think we need `l--` after the assignment.